File: boundary/boundary-solver.py
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm


Nx = 160000
Ny = 80
delta0 = 0.05
L = 0.4 

x = np.linspace(0, L, Nx)
#y = delta0 * (np.tanh(2 * np.linspace(0, 1, Ny)) / np.tanh(2))
y = np.linspace(0, delta0, Ny)
dx = x[1] - x[0]
print(dx)
dy = y[1] - y[0]
print(dy**2)

rho = 1
mu = 1.95*1e-5
nu = mu / rho
dp_dx = 0#0.01*1e5

u = np.zeros((Nx, Ny))
v = np.zeros((Nx, Ny))
nut = np.zeros((Nx, Ny))
mut = np.zeros((Nx, Ny))
kappa = 0.41

count = 0

for j in range(Ny):
    eta = y[j]/delta0
    u[0, j] = 1.5 * eta - 0.5 * eta**3 if eta<1 else 1


for i in tqdm(range(1, Nx)):
    #delta = delta0 + 0.005 * x[i]     # Placeholder layer growth relation
    delta = delta0 + np.sqrt(x[i]*mu*rho)
    count+=1
    #print(count)


    eta = y / delta
    nut[i-1, :] = np.where(eta < 1, 0.01 * (eta*(1-eta)), 0)
    mut[i-1, :] = rho * nut[i-1, :]

    for j in range(1, Ny-1):
        du_dy = (u[i-1, j+1] - u[i-1, j-1]) / (2*dy)

        d2u_dy2 = (u[i-1, j+1] - 2*u[i-1, j] + u[i-1, j-1]) / (dy)**2

        visc_term = (mu + mut[i-1, j])*d2u_dy2 + (
            (mut[i-1, j+1] - mut[i-1, j-1])/(2*dy)
            ) * du_dy
        
        du_dx = (
            (visc_term - dp_dx)/rho - v[i-1, j]*du_dy
            )/u[i-1, j] if abs(u[i-1, j]) > 1e-8 else 0
        
        u[i, j] = u[i-1, j] + dx * du_dx

    u[i, 0] = 0             # No slip condition
    u[i, -1] = 1

    v[i, -1] = 0
    for j in reversed(range(1, Ny)):
        du_dx = (u[i, j] - u[i-1, j]) / dx
        v[i, j-1] = v[i, j] - dy * du_dx

dudy_w = (u[100, 1] - u[100, 0]) / dy
tau_w = mu*dudy_w
u_tau = np.sqrt(tau_w / rho)

y_plus = y*u_tau / nu
x_plus = u[100, :] / u_tau


plt.figure()
plt.plot(u[0, :], y)
plt.show(block=False)
plt.plot(u[10, :], y)
plt.show(block=False)
plt.plot(u[100, :], y)
plt.show(block=False)
plt.plot(u[1000, :], y)
plt.show(block=False)
plt.plot(u[10000, :], y, label='1')
plt.show(block=False)
plt.plot(u[40000, :], y, label='4')
plt.show(block=False)
plt.plot(u[80000-1, :], y, label='8')
plt.show(block=False)
plt.plot(u[160000-1, :], y, label='16')
plt.title("Boundary Layer 40y (with y and delta functions + 0.01bar)")
plt.ylabel('y position')
plt.xlabel('Velocity distribution')
plt.legend()
plt.show(block=False)

plt.figure()
plt.plot(y_plus, x_plus)
plt.xscale('log')
plt.title('Plot for x=100')
plt.xlabel('y+')
plt.ylabel('u+')
plt.show()

X, Y = np.meshgrid(x, y)
plt.figure(figsize=(8, 4))
plt.contourf(X, Y, u.T, levels=50, cmap="viridis")
plt.colorbar(label="u [m/s]")
plt.xlabel("x [m]")
plt.ylabel("y [m]")
plt.title("Velocity field u(x, y)")
plt.tight_layout()
plt.show()

File: boundary/boundary-solver.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <fstream>

using namespace std;

int main() {
    const int Nx = 160000*4;
    const int Ny = 120;
    const double delta0 = 0.05;
    const double L = 0.02*2;
    const double rho = 1;
    const double mu = 1.95e-5;
    const double nu = mu / rho;
    const double dp_dx = 0.0;
    const double U_x = 1;
    const double kappa = 0.41;

    const int yp_iter = 160000;

    int count = 0;

    vector<double> x(Nx);
    vector<double> y(Ny);
    for (int i = 0; i < Nx; ++i) {
        x[i] = L * i / (Nx-1);
    }

    for (int j = 0; j < Ny; ++j) {
        y[j] = pow(1.0 * j / (Ny-1), 1.3) * delta0;
    }

    double dx = x[1] - x[0];
    vector<double> dy(Ny);
    dy[0] = y[1] - y[0];
    dy[Ny-1] = y[Ny-1] - y[Ny-2];

    for (int j=1; j < Ny-1; ++j) {
        dy[j] = (y[j+1] - y[j-1]) / 2;
    }

    cout << "dx = " << dx << endl;
    cout << "dy = " << pow(dy[0], 2) << endl;

    vector<vector<double>> u(Nx, vector<double>(Ny, 0));
    vector<vector<double>> v(Nx, vector<double>(Ny, 0));
    vector<vector<double>> nut(Nx, vector<double>(Ny, 0));
    vector<vector<double>> mut(Nx, vector<double>(Ny, 0));
    vector<vector<double>> tau_v(Nx, vector<double>(Ny, 0));
    vector<vector<double>> tau_t(Nx, vector<double>(Ny, 0));
    vector<vector<double>> dudy(Nx, vector<double>(Ny, 0));

    for (int j = 0; j < Ny; ++j) {
        double eta = y[j] / delta0;
        u[0][j] = (eta < 1.0) ? (1.5 * eta - 0.5 * pow(eta, 3)) : 1.0;
    }

    dudy[0][0] = (u[0][1] - u[0][0]) / dy[0];
    dudy[0][Ny-1] = (u[0][Ny-1] - u[0][Ny-2]) / dy[Ny-1];

    for (int j = 1; j < Ny-1; ++j) {
        dudy[0][j] = (u[0][j+1] - u[0][j-1]) / (2*dy[j]);
    }

    for (int i = 1; i < Nx; ++i) {

        for (int j = 0; j < Ny; ++j) {
            nut[i-1][j] = pow((kappa * y[j]), 2) * fabs(dudy[i-1][j]);
            mut[i-1][j] = rho * nut[i-1][j];
        }

        for (int j = 1; j < Ny-1; ++j) {
            double du_dy = (u[i-1][j+1] - u[i-1][j-1]) / (2*dy[j]);
            dudy[i][j] = du_dy;

            double d2u_dy2 = (u[i-1][j+1] - 2 * u[i-1][j] + u[i-1][j-1]) / pow(dy[j], 2);

            double visc_term = (mu + mut[i-1][j]) * d2u_dy2 + (
                (mut[i-1][j+1] - mut[i-1][j-1]) / (2*dy[j])
            ) * du_dy;

            tau_v[i-1][j] = mu * du_dy;
            tau_t[i-1][j] = mut[i-1][j]*du_dy;

            double du_dx = (fabs(u[i-1][j]) > 1e-8) ? (
                ((visc_term - dp_dx) / rho - v[i-1][j] * du_dy) / u[i-1][j]
            ) : 0;

            u[i][j] = u[i-1][j] + dx * du_dx;
        }

        dudy[i][0] = (u[i][1] - u[i][0]) / dy[0];       // Wall gradient

        u[i][0] = 0;        // No slip condition
        u[i][Ny-1] = U_x;     // Outer stream flow condition U(x) = 1

        v[i][Ny-1] = 0;

        for (int j = Ny-1; j > 0; --j) {
            double du_dx = (u[i][j] - u[i-1][j]) / dx;
            v[i][j-1] = v[i][j] - dy[j] * du_dx;
        }
    }

    //double dudy_w = (u[yp_iter][1] - u[yp_iter][0]) / dy[0];
    double tau_w = mu * dudy[yp_iter][0];
    double cf = (2 * tau_w) / (rho * pow(U_x, 2));
    cout << "dudy = " << dudy[yp_iter][0] << endl;
    cout << "cf = " << cf << endl;
    double u_tau = sqrt(tau_w / rho);
    double delta_v = nu / u_tau;
    cout << "delta_v = " << delta_v << endl;
    
    vector<double> y_plus(Ny);
    vector<vector<double>> u_plus(Nx, vector<double>(Ny, 0.0));

//    for (int j = 0; j < Ny; ++j) {
//        y_plus[j] = y[j] * u_tau / nu;
//    }

//    for (int i = 0; i < Nx; ++i) {          // Add to main loop for greater efficiency
//        for (int j = 0; j < Ny; ++j) {
//            u_plus[i][j] = u[i][j] / u_tau;
//        }
//    }

    ofstream outfile("./boundary-layer.csv");
    outfile << "y" << "," << "u[0]" << "," << "u[10000]" << "," << "u[80000]" << ",";
    outfile << "u[160000]" << "," << "u[320000]" << "," << "u[640000]" << ",";
    outfile << "y+" << "," << "u+" << "," << "tau_v[40000]" << "," << "tau_t[40000]" << ",";
    outfile << "tau_v[640000]" << "," << "tau_t[640000]" << endl;
    for (int j = 0; j<Ny; ++j) {
        y_plus[j] = y[j] * u_tau / nu;
        outfile << y[j] << "," << u[0][j] << "," << u[10000][j] << "," << u[80000][j] << ",";
        outfile << u[160000][j] << "," << u[320000-1][j] << "," << u[640000-1][j] << ",";
        outfile << y_plus[j] << "," << 0 << "," << tau_v[40000][j] << "," << tau_t[40000][j] << ",";
        outfile << tau_v[640000-1][j] << "," << tau_t[640000-1][j] << endl;
    }
    outfile.close();

    cout << "Simulation complete. Data saved to boundary-layer.csv" << endl;

    //system("pause");
}